### ISSUE TITLE
Dash (-) not supported in resource name

### ISSUE THREAD

**Issue by prshreshtha** [holidayextras/jsonapi-server#359]
The resource and attribute names are checked against the regex `/^[A-Za-z0-9_]*$/`. Is it possible to also support `-`, i.e. `resource-name-is-like-this` format? It could just be as easy as changing it to `/^[A-Za-z0-9_\-]*$/`.

**Comment by pmcnr-hx** [comment]
I will have to check the spec to see if `-` is allowed. I'll come back to you on this one.

**Comment by prshreshtha** [comment]
We see [here](http://jsonapi.org/format/#document-resource-object-identification):
> The values of type members MUST adhere to the same constraints as member names.

Which says
> The following “globally allowed characters” MAY be used anywhere in a member name:
> * U+0061 to U+007A, “a-z”
> * U+0041 to U+005A, “A-Z”
> * U+0030 to U+0039, “0-9”
> * U+0080 and above (non-ASCII Unicode characters; not recommended, not URL safe)

> Additionally, the following characters are allowed in member names, except as the first or last character:
> * U+002D HYPHEN-MINUS, “-“
> * U+005F LOW LINE, “_”
> * U+0020 SPACE, “ “ (not recommended, not URL safe)

The requirements in the spec are pretty clear and should be simple to extract into a function instead of a regex test. Let me know what you think.

**Comment by pmcnr-hx** [comment]
You are correct @prshreshtha. You're welcome to submit a PR for review and eventual merging.

**Comment by prshreshtha** [comment]
Sounds good. I'll make one soon.

**Comment by prshreshtha** [comment]
It looks like GraphQL does not support dasherized field names. Any thoughts @pmcnr-hx?

**Comment by paparomeo** [comment]
Ah yes. I seemed to recall there was a reason not to support it. One options is to replace the `-` with `_` on the fly. The other one is to disable GraphQL support if you prefer to use just use JSON:API with `-` on the names. I don't think GraphQL support should prevent us from fully supporting the JSON:API spec. Which option do you prefer?

**Comment by prshreshtha** [comment]
@paparomeo there are trade-offs here. I don't think replacing `-` with `_` will work without breaking the API since `my-resource` and `my_resource` will collide. I think the ability to disable GraphQL support to take full advantage of JSON:API is the way to go and document the stricter requirements when using GraphQL. This definitely needs some research.

**Comment by paparomeo** [comment]
Thanks for the investigation @prshreshtha. I should have some time soon to look into adding support for disabling GraphQL.

**Comment by queenvictoria** [comment]
You can already do this. `graphiql: false` in the options.

**Comment by prshreshtha** [comment]
@queenvictoria You are correct that #256 does cover part of this issue. It is a pre-requisite to supporting dasherized names. But I don't see any work done for it.
